### PR TITLE
Rordev 117

### PR DESCRIPTION
**Pull request by clutroth**

### CODE REVIEW

**Comment by coutoPL** [comment]
@clutroth LoggingContext can be determined at ROR core instantiation level, so maybe we can pass it in constructor, not in methods?

**Comment by clutroth** [comment]
> @clutroth LoggingContext can be determined at ROR core instantiation level, so maybe we can pass it in constructors, not in methods?

Thanks, mess starts in `IndexLevelActionFilter`.  `LoggingContext` is used by `CurrentUserMetadataRequestHandler` and `RegularRequestHandler`. I didn't found simpler solution. Do you have better idea for this?

I had problem with `Acl` creation, but solved it.

**Comment by coutoPL** [comment]
> > @clutroth LoggingContext can be determined at ROR core instantiation level, so maybe we can pass it in constructors, not in methods?
> 
> Thanks, mess starts in `IndexLevelActionFilter`. `LoggingContext` is used by `CurrentUserMetadataRequestHandler` and `RegularRequestHandler`. I didn't found simpler solution. Do you have better idea for this?
> 
> I had problem with `Acl` creation, but solved it.

I see it as follows:
* all logs which should be obfuscated are produced by core (especially AclLoggingDecorator and rules)
* ROR core is created in RawRorConfigBasedCoreFactory
* obfuscated headers are loaded from config 
* so when we instantiate core, we can pass LoggingContext to ACL logging decorator constructor and to rules which need it.

Does it make sense?

File: core/src/main/scala/tech/beshu/ror/accesscontrol/logging/AccessControlLoggingDecorator.scala [hunk 1]
```diff
@@ -115,7 +118,7 @@ class AccessControlLoggingDecorator(val underlying: AccessControl, auditingTool:
 
 object AccessControlLoggingDecorator {
 
-  private implicit val responseContextShow: Show[ResponseContext] = {
+  private implicit def responseContextShow(implicit headerShow:Show[Header]): Show[ResponseContext] = {
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/logging/AccessControlLoggingDecorator.scala]: why not to pass here whole LoggingContext? In future we'd like to add more data to LoggingContext, so there will be no need to change signature of this method.

File: core/src/main/scala/tech/beshu/ror/accesscontrol/logging/LoggingContext.scala [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package tech.beshu.ror.accesscontrol.logging
+
+import cats.Show
+import tech.beshu.ror.accesscontrol.domain.Header
+
+final class LoggingContext(implicit val showHeader:Show[Header])
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/logging/LoggingContext.scala]: IMO we should store in LoggingContext data, not typeclasses. I mean, IMO it's better to defined is like this:
`final class LoggingContext(obfuscatedHeaders: Set[Header.Name])`

Then you can use this class not only to stringify headers, but also to filter them, etc.

File: core/src/main/scala/tech/beshu/ror/accesscontrol/logging/LoggingContextFactory.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package tech.beshu.ror.accesscontrol.logging
+
+import cats.Show
+import cats.implicits._
+import tech.beshu.ror.accesscontrol.domain.Header
+import tech.beshu.ror.accesscontrol.show.logs._
+
+object LoggingContextFactory {
+  def create(obfuscatedHeaders: Option[ObfuscatedHeaders]): LoggingContext = {
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/logging/LoggingContextFactory.scala]: I think we don't need ObfuscatedHeaders class. I know that there are different approaches according to work with config files:
* parse to intermediate objects and then to core classes
* parse directly to core classes
* [and others]

Your approach is the first one. But in ROR we decided to use 2nd one and we should follow this convention. I know sometimes there are real hard issues to solve using it, but consistency in code base is the most important thing,

File: core/src/main/scala/tech/beshu/ror/accesscontrol/logging/ObfuscatedHeaders.scala [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package tech.beshu.ror.accesscontrol.logging
+
+import tech.beshu.ror.accesscontrol.domain.Header
+
+final case class ObfuscatedHeaders(headers:Set[Header.Name])
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/logging/ObfuscatedHeaders.scala]: to remove

File: core/src/main/scala/tech/beshu/ror/accesscontrol/ops.scala [hunk 5]
```diff
@@ -133,7 +129,7 @@ object show {
     implicit val dnShow: Show[Dn] = Show.show(_.value.value)
     implicit val envNameShow: Show[EnvVarName] = Show.show(_.value.value)
     implicit val propNameShow: Show[PropName] = Show.show(_.value.value)
-    implicit val blockContextShow: Show[BlockContext] = Show.show { bc =>
+    implicit def blockContextShow(implicit showHeader:Show[Header]): Show[BlockContext] = Show.show { bc =>
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/ops.scala]: please use LoggingContext here

File: core/src/main/scala/tech/beshu/ror/accesscontrol/ops.scala [hunk 6]
```diff
@@ -171,7 +167,7 @@ object show {
         }
       }"
     }
-    implicit val historyShow: Show[History] = Show.show { h =>
+    implicit def historyShow(implicit headerShow:Show[Header]): Show[History] = Show.show { h =>
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/ops.scala]: please use LoggingContext here

File: core/src/main/scala/tech/beshu/ror/boot/ReadonlyRest.scala [hunk 7]
```diff
@@ -50,15 +51,15 @@ import scala.util.Success
...
+  val blockingScheduler: Scheduler = Scheduler.io("blocking-index-content-provider")
 
   override protected val envVarsProvider: EnvVarsProvider = OsEnvVarsProvider
   override protected implicit val clock: Clock = Clock.systemUTC()
 
   override protected val coreFactory: CoreFactory = {
     implicit val uuidProvider: UuidProvider = JavaUuidProvider
     implicit val propertiesProvider: PropertiesProvider = JvmPropertiesProvider
-    implicit val _ = envVarsProvider
+    implicit val _envVarsProvider: EnvVarsProvider = envVarsProvider
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply core/src/main/scala/tech/beshu/ror/boot/ReadonlyRest.scala]: please remove _

File: core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala [hunk 8]
```diff
@@ -26,28 +26,33 @@ import tech.beshu.ror.accesscontrol.blocks.Block.ExecutionResult.{Matched, Misma
...
-
 class Block(val name: Name,
             val policy: Policy,
             val verbosity: Verbosity,
             val rules: NonEmptyList[Rule])
   extends Logging {
 
-  def execute(requestContext: RequestContext): BlockResultWithHistory = {
+  def execute(requestContext: RequestContext)
+             (implicit loggingContext: LoggingContext): BlockResultWithHistory = {
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala]: no fix, no comment

**Comment by coutoPL** [comment]
the are merge conflicts to be resolved

File: core/src/main/scala/tech/beshu/ror/accesscontrol/acl/AccessControlList.scala [hunk 9]
```diff
@@ -27,12 +27,14 @@ import tech.beshu.ror.accesscontrol.blocks.Block.{ExecutionResult, History, Poli
...
 import tech.beshu.ror.accesscontrol.domain.Group
+import tech.beshu.ror.accesscontrol.logging.LoggingContext
 import tech.beshu.ror.accesscontrol.orders.forbiddenByMismatchedCauseOrder
 import tech.beshu.ror.accesscontrol.request.RequestContext
 import tech.beshu.ror.accesscontrol.request.RequestContextOps._
 import tech.beshu.ror.utils.uniquelist.UniqueList
 
-class Acl(val blocks: NonEmptyList[Block])
+class AccessControlList(val blocks: NonEmptyList[Block])
+                       (implicit val loggingContext: LoggingContext)
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/acl/AccessControlList.scala]: I thought that only logging decorator needs it, doesn't it?

File: core/src/main/scala/tech/beshu/ror/accesscontrol/logging/ObfuscatedHeaderShowFactory.scala [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package tech.beshu.ror.accesscontrol.logging
+
+import cats.Show
+import cats.implicits._
+import tech.beshu.ror.accesscontrol.domain.Header
+import tech.beshu.ror.accesscontrol.show.logs._
+
+object ObfuscatedHeaderShowFactory {
+  def create(obfuscatedHeaders: Set[Header.Name]): Show[Header] = {
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/logging/ObfuscatedHeaderShowFactory.scala]: please move it close to other Show typeclass implementations (tech.beshu.ror.accesscontrol.show.logs)

File: core/src/main/scala/tech/beshu/ror/accesscontrol/acl/AccessControlList.scala (hunk 9 above)
Comment (clutroth) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/acl/AccessControlList.scala]: No, we're writing logs in other classes to. `Block` for instance. L135, L153

File: core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (clutroth) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala]: obraz

[the remark this answers is not in the document]
Comment (clutroth) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala]: comment

[the remark this answers is not in the document]
Comment (clutroth) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala]: Oh, ok. I didn't click "Finish review", so comments are pending. Not everybody have to be fluent with computers.

File: core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala [hunk 11]
```diff
@@ -38,16 +41,20 @@ class Block(val name: Name,
...
+  def executeUserMetadataRuleOnly(requestContext: RequestContext)
+                                 (implicit loggingContext: LoggingContext): BlockResultWithHistory = {
     processRules(rules.collect { case r: UserMetadataRelatedRule => r }, requestContext)
   }
 
   private def processRules(selectedRules: List[Rule],
-                           requestContext: RequestContext): BlockResultWithHistory = {
+                           requestContext: RequestContext)
+                          (implicit loggingContext: LoggingContext): BlockResultWithHistory = {
+    implicit val showHeader: Show[Header] = ObfuscatedHeaderShowFactory.create(loggingContext.obfuscatedHeaders)
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/Block.scala]: please move creation of show instance close to Show[RequestContext] instance - seems that this code is only interested in using the showHeader

File: core/src/test/scala/tech/beshu/ror/unit/logging/ObfuscatedHeaderShowFactoryTest.scala [hunk 12]
```diff
@@ -0,0 +1,38 @@
+package tech.beshu.ror.unit.logging
+
+import eu.timepit.refined.types.string.NonEmptyString
+import org.scalatest.prop.TableDrivenPropertyChecks
+import org.scalatest.{Matchers, WordSpec}
+import tech.beshu.ror.accesscontrol.domain.Header
+import tech.beshu.ror.accesscontrol.show.ObfuscatedHeaderShowFactory
+
+class ObfuscatedHeaderShowFactoryTest
```
Comment (coutoPL) [inline core/src/test/scala/tech/beshu/ror/unit/logging/ObfuscatedHeaderShowFactoryTest.scala]: remember to change name of the test too

**Comment by sscarduzio** [comment]
@clutroth thanks for the PR! Could you please commit the final changes within today?

**Comment by clutroth** [comment]
Yes.
<http://www.czegoszukasz.org>Pozdrawiam
Wojciech Decker

Czego Szukasz <http://www.czegoszukasz.org>?


pt., 20 wrz 2019 o 07:47 Simone Scarduzio <notifications@github.com>
napisał(a):

**Comment by coutoPL** [comment]
LGTM. Thanks!

**Pull request merged**